src/charm.py

```diff
@@ -57,44 +68,93 @@ def _on_install(self, _):
         """Handler for the install event during which we will update the K8s service."""
         self._patch_k8s_service()
 
-    def _on_config_changed(self, _):
-        if self._stored.provider_ready:
-            self.unit.status = ActiveStatus()
+    def _on_config_changed(self, event):
+        self._configure(event)
 
     def _on_loki_pebble_ready(self, event):
-        """Define and start a workload using the Pebble API."""
-        # Get a reference the container attribute on the PebbleReadyEvent
-        container = event.workload
-        # Define an initial Pebble layer configuration
+        self._configure(event)
+
+    def _on_alertmanager_change(self, event):
+        self._configure(event)
+
+    def _configure(self, event):
+        """Configure Loki"""
+        restart = False
+
+        if not self._container.can_connect():
+            self.unit.status = WaitingStatus("Waiting for Pebble ready")
+            return False
+
+        current_layer = self._container.get_plan().services
+        new_layer = self._build_pebble_layer
+
+        if current_layer != new_layer:
+            restart = True
+
+        config = self._loki_config()
+
+        try:
+            if yaml.safe_load(self._stored.config) != config:
```

Comment:
I am not sure why you bother to check if the current config is different from the saved one. Why not just always push the current config and simplify your code flow ?

Reply:
I am checking whether the current config is different from the saved one to restart the container by setting `restart = True`

Do you think there is a better approach here?

Reply:
See #14